packages/openapi-typescript/test/transform/components-object.test.ts

```diff
@@ -464,6 +464,37 @@ describe("transformComponentsObject", () => {
         options: { ...DEFAULT_OPTIONS, excludeDeprecated: true },
       },
     ],
+    [
+      "options > rootTypes: true",
+      {
+        given: {
+          schemas: {
+            SomeType: {
+              type: "object",
+              properties: {
+                name: { type: "string" },
+                url: { type: "string" },
+              },
+            },
+          },
+        },
+        want: `{
+    schemas: {
+        SomeType: {
+            name?: string;
+            url?: string;
+        };
+    };
+    responses: never;
+    parameters: never;
+    requestBodies: never;
+    headers: never;
+    pathItems: never;
+}
+export type SomeTypeSchema = components['schemas']['SomeType'];`,
+        options: { ...DEFAULT_OPTIONS, rootTypes: true },
```

Comment:
So this is a great start to testing this, and this is exactly what we need! But we will need a lot more tests to handle all of the scenarios [from my original requirements](https://github.com/openapi-ts/openapi-typescript/pull/1260#issuecomment-1658694431):

1. **Conflicts**: test that `#/components/schemas/schemaFoo` and `#/components/schemas/Schemafoo` don’t conflict (and many combinations including hyphens and invalid JS characters like leading numbers!). We’ll need to handle:
  - Case sensitivity (this is handled by the large object, and this is intentional)
  - Invalid characters: `-` and `.` are the most common. Beyond that IMO not too important.
  - Leading numbers (e.g. `1schema`)
2. **ALL of components object, not just `#/components/schemas`:
  - `#/components/schemas`
  - `#/components/responses`
  - `#/components/parameters`
  - `#/components/requestBodies`

We will need all of this to happen in the same PR, because it’s very likely that supporting all this could lead to refactoring and churn based on the implementation. As-is I like the path you’re on! But we need to have all of these present and tested before we merge this.

Reply:
Addressed point 1 with [d40a7fd](https://github.com/openapi-ts/openapi-typescript/pull/1876/commits/d40a7fd9e852d61c62c68c116b8886ea4bf9ed84). All conflicting names are suffixed by a `_2`, `_3`, etc. as per the original discussion in the previous PR you linked. This behavior is up for discussion. Not sure this is a great developer experience, but at the same time don't feel it warrants an error given the OpenAPI spec permits these things. Given that, figured this should be the compliant portion and work around the limitations of JS to comply to the OpenAPI spec as closely as possible. Maybe possible to just print warnings about conflicts and automatic suffixing if detected as a happy compromise?

Filled out the test to cover the other components properties in [7edf333](https://github.com/openapi-ts/openapi-typescript/pull/1876/commits/7edf3336fff3929da0a2d78583de99da357ef522) and tested for conflicts across them in [f64e0e6](https://github.com/openapi-ts/openapi-typescript/pull/1876/commits/f64e0e6b98199654a12cc2596600345fad1fceab)

Reply:
> Given that, figured this should be the compliant portion and work around the limitations of JS to comply to the OpenAPI spec as closely as possible.

+1

> Maybe possible to just print warnings about conflicts and automatic suffixing if detected as a happy compromise?

Yeah I’m open to that, but don’t feel strongly. I think developers will realize if they’re importing the wrong things quickly; the main thing to handle was just types disappearing / not exporting from `rootTypes` at all would have people raising bug reports.